storage/perfschema/table_malloc_stats.cc

```diff
@@ -0,0 +1,269 @@
+  jemalloc_mallctl("epoch", nullptr, 0, &epoch, sz);
+
+  for (unsigned i = 0; i < NUM_SUMMARY_STAT; i++)
+    jemalloc_mallctl(summary_stat_type[i], &m_row.stat[i], &sz, nullptr, 0);
+
+  m_row_exists = true;
+}
+
+int table_malloc_stats_totals::read_row_values(TABLE *table, unsigned char *,
+                                                Field **fields, bool read_all) {
+  Field *f;
+
+  if (unlikely(!m_row_exists)) return HA_ERR_RECORD_DELETED;
+
+  /* Set the null bits */
+  DBUG_ASSERT(table->s->null_bytes == 1);
+
+  for (; (f = *fields); fields++) {
+    if (read_all || bitmap_is_set(table->read_set, f->field_index())) {
+      DBUG_ASSERT(f->field_index() < NUM_SUMMARY_STAT);
+      set_field_ulonglong(f, m_row.stat[f->field_index()]);
+    }  // if
+  }    // for
+
+  return 0;
+}
+
+THR_LOCK table_malloc_stats::m_table_lock;
+
+static const char *row_type[] = {"small", "large", "huge"};
+#define NUM_ROWS (sizeof(row_type) / sizeof(row_type[0]))
+static const char *stat_type[NUM_STAT] = {"allocated", "nmalloc", "ndmalloc",
+                                          "nrequests"};
+
+Plugin_table table_malloc_stats::m_table_def(
+    /* Schema name */
+    "performance_schema",
+    /* Name */
+    "malloc_stats",
+    /* Definition */
+    " TYPE char(8) not null,\n"
+    " ALLOCATED bigint(20) unsigned not null,\n"
+    " NMALLOC bigint(20) unsigned not null,\n"
+    " NDALLOC bigint(20) unsigned not null,\n"
+    " NREQUESTS bigint(20) unsigned not null\n",
+    /* Options */
+    " ENGINE=PERFORMANCE_SCHEMA",
+    /* Tablespace */
+    nullptr);
+
+PFS_engine_table_share table_malloc_stats::m_share = {
+    &pfs_readonly_acl,
+    table_malloc_stats::create,
+    nullptr, /* write_row */
+    nullptr, /* delete_all_rows */
+    table_malloc_stats::get_row_count,
+    sizeof(PFS_simple_index),
+    &m_table_lock,
+    &m_table_def,
+    false, /* perpetual */
+    PFS_engine_table_proxy(),
+    {0},
+    false /* m_in_purgatory */
+};
+
+PFS_engine_table *table_malloc_stats::create(PFS_engine_table_share *) {
+  return new table_malloc_stats();
+}
+
+ha_rows table_malloc_stats::get_row_count() { return 5; }
+
+table_malloc_stats::table_malloc_stats()
+    : PFS_engine_table(&m_share, &m_pos),
+      m_row_exists(false),
+      m_pos(0),
+      m_next_pos(0) {}
+
+void table_malloc_stats::reset_position(void) {
+  m_pos.m_index = 0;
+  m_next_pos.m_index = 0;
+}
+
+int table_malloc_stats::rnd_next(void) {
+  m_pos.set_at(&m_next_pos);
+
+  if (m_pos.m_index < NUM_ROWS) {
+    make_row(m_pos.m_index);
+    m_next_pos.set_after(&m_pos);
+    return 0;
+  }
+
+  return HA_ERR_END_OF_FILE;
+}
+
+int table_malloc_stats::rnd_pos(const void *pos) {
+  set_position(pos);
+
+  if (m_pos.m_index < NUM_ROWS) {
+    make_row(m_pos.m_index);
+    return 0;
+  }
+
+  return HA_ERR_RECORD_DELETED;
+}
+
+void table_malloc_stats::make_row(int index) {
+  char s[64];
+  size_t sz = sizeof(epoch);
+
+  memset(&m_row, 0, sizeof(m_row));
+
+  epoch++;
```

Comment:
Is the access protected by mutex? What if different threads are accessing this table?

Reply:
None of the other methods (`rnd_next`, etc) of the rest of the `perfschema` tables are protected by a mutex, please, are you talking of the access to `epoch`?

Reply:
Thanks for the clarification. I think this is not protected by the static mutex of the table class. I will turn it into atomic.